### PR TITLE
ppx_repr: add support for meta-deriving plugins

### PR DESCRIPTION
**Pull request by CraigFe**
This feature adds support for "meta" deriving plugins for `ppx_repr`. This is a
hook for injecting various specialisations of generic operations whenever a
`Repr.t` value is derived. The idea is to get rid of the following boilerplate:

```ocaml
type value = (* ... *) [@@deriving repr]
let equal_value = Repr.(unstage (of_bin_string value_t))
let value_of_bin_string = Repr.(unstage (of_bin_string value_t))
let value_decode_bin = Repr.(unstage (decode_bin value_t))
```

The above can now be written as:

```ocaml
type value = V.t [@@deriving repr ~equal ~decode_bin ~of_bin_string]
```

which expands out to the long form variant.

### CODE REVIEW

**Comment by Ngoguey42** [comment]
Have you considered the solution where the default behaviour is exporting all operations?

**Comment by CraigFe** [comment]
I did think of it, but am not sure it's a good idea to silently introduce structural definitions of things (thinking of our recent `compare` / `equal` bugs in Tree). I'm also concerned that would add a lot to the initialise time of libraries linking against Irmin, but I haven't tested it.

**Comment by Ngoguey42** [comment]
Agreed

File: src/ppx_repr/lib/meta_deriving.ml [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+      create "pp_dump"
+        ~impl:(fun loc t -> [%expr Repr.pp_dump [%e t]])
+        ~intf:(fun loc t -> [%type: Stdlib.Format.formatter -> [%t t] -> unit]);
+      create "random"
+        ~impl:(fun loc t -> [%expr Repr.unstage (Repr.random [%e t])])
+        ~intf:(fun loc t -> [%type: unit -> [%t t]]);
+      create "to_bin_string" ~type_name:`before
+        ~impl:(fun loc t -> [%expr Repr.unstage (Repr.to_bin_string [%e t])])
+        ~intf:(fun loc t -> [%type: [%t t] -> string]);
+      create "of_bin_string" ~type_name:`before
```
Comment (mattiasdrp) [inline src/ppx_repr/lib/meta_deriving.ml:65]: Is there a reason why we'd want automatic deriving to prefix or suffix the operation?

File: src/repr/type_intf.ml [hunk 2]
```diff
@@ -376,7 +376,7 @@ module type DSL = sig
   val compare : 'a t -> 'a compare
   (** [compare t] compares values of type [t]. *)
 
-  type 'a pp = 'a Fmt.t
+  type 'a pp = Format.formatter -> 'a -> unit
```
Comment (mattiasdrp) [inline src/repr/type_intf.ml:379]: What is the reason behind this change? Readability?

File: src/ppx_repr/lib/engine.ml [hunk 3]
```diff
@@ -22,6 +22,8 @@ let map_lident f = function
   | Ldot (l, s) -> Ldot (l, f s)
   | Lident s -> Lident (f s)
 
+let repr_name_of_type_name = function "t" -> "t" | x -> x ^ "_t"
```
Comment (mattiasdrp) [inline src/ppx_repr/lib/engine.ml:25]: Regarding this, I was wondering if it would be possible, instead of deciding that `t` is the _default_ type to add a `~main` option that tells the deriver we don't want representatives and functions associated to this type to be prefixed or suffixed. We would enforce the existence of a single `main` type in a module and give more control over what the deriver will produce.

File: src/ppx_repr/lib/meta_deriving.ml (hunk 1 above)
Comment (CraigFe) [reply src/ppx_repr/lib/meta_deriving.ml:65]: Mostly just interop with the existing conventions used in Irmin:

```ocaml
type value
val pp_value : ...            (* prefix the operation *)
val value_of_bin_string : ... (* suffix the operation *)
```

These seem sensible to me, so I chose to bake it in here.

`ppx_deriving` has a similar feature in [its conventions](https://github.com/ocaml-ppx/ppx_deriving#plugin-conventions) (although it also provides an option for the user to override the convention, which I think is a bit needless).

**Review by Ngoguey42 — APPROVED** [review]
LGTM

File: src/ppx_repr/lib/engine.ml (hunk 3 above)
Comment (CraigFe) [reply src/ppx_repr/lib/engine.ml:26]: Good point. I'd like to have some particular use-case in mind before generalising in that way, but in principle it sounds worthwhile.

**Comment by CraigFe** [comment]
Thanks both for the reviews :-)
